diff --git a/components/partition_table/project_include.cmake b/components/partition_table/project_include.cmake
index 7778af1155..f9b5138716 100644
--- a/components/partition_table/project_include.cmake
+++ b/components/partition_table/project_include.cmake
@@ -13,10 +13,14 @@ if(NOT BOOTLOADER_BUILD)
         if(IS_ABSOLUTE ${CONFIG_PARTITION_TABLE_FILENAME})
             set(PARTITION_CSV_PATH "${CONFIG_PARTITION_TABLE_FILENAME}")
         else()
+            if(NOT CONFIG_PARTITION_TABLE_BASE_DIR)
+                set(CONFIG_PARTITION_TABLE_BASE_DIR "${project_dir}")
+            endif()
+
             idf_build_get_property(project_dir PROJECT_DIR)
             # Custom filename expands any path relative to the project
             get_filename_component(PARTITION_CSV_PATH "${CONFIG_PARTITION_TABLE_FILENAME}"
-                                    ABSOLUTE BASE_DIR "${project_dir}")
+                                    ABSOLUTE BASE_DIR "${CONFIG_PARTITION_TABLE_BASE_DIR}")
         endif()
 
         if(NOT EXISTS "${PARTITION_CSV_PATH}")
