diff --git a/tools/cmake/gdbinit.cmake b/tools/cmake/gdbinit.cmake
index d4b8775631..53537cfe8a 100644
--- a/tools/cmake/gdbinit.cmake
+++ b/tools/cmake/gdbinit.cmake
@@ -30,14 +30,14 @@ function(__generate_gdbinit)
 
     # Define paths
     set(gdbinit_dir ${BUILD_DIR}/gdbinit)
-    set(gdbinit_rom_in_path ${gdbinit_dir}/rom.gdbinit.in)
-    set(gdbinit_rom_path ${gdbinit_dir}/rom.gdbinit)
+    # set(gdbinit_rom_in_path ${gdbinit_dir}/rom.gdbinit.in)
+    # set(gdbinit_rom_path ${gdbinit_dir}/rom.gdbinit)
     set(symbols_gdbinit_path ${gdbinit_dir}/symbols)
     set(py_extensions_gdbinit_path ${gdbinit_dir}/py_extensions)
     set(connect_gdbinit_path ${gdbinit_dir}/connect)
     idf_build_get_property(PROJECT_EXECUTABLE EXECUTABLE)
     set(application_elf ${BUILD_DIR}/${PROJECT_EXECUTABLE})
-    file(TO_CMAKE_PATH $ENV{ESP_ROM_ELF_DIR} ESP_ROM_ELF_DIR)
+    # file(TO_CMAKE_PATH $ENV{ESP_ROM_ELF_DIR} ESP_ROM_ELF_DIR)
 
     file(MAKE_DIRECTORY ${gdbinit_dir})
 
@@ -45,6 +45,7 @@ function(__generate_gdbinit)
     if(CONFIG_IDF_TARGET_LINUX)
         set(rom_symbols)
     else()
+    if(0)
         execute_process(
             COMMAND python "${idf_path}/components/esp_rom/gen_gdbinit.py" ${IDF_TARGET}
             OUTPUT_VARIABLE rom_symbols
@@ -55,6 +56,7 @@ function(__generate_gdbinit)
             message(WARNING "Error while generating esp_rom gdbinit")
         endif()
     endif()
+    endif()
 
     # Check if bootloader ELF is defined and set symbol-file accordingly
     if(DEFINED BOOTLOADER_ELF_FILE)
