diff --git a/components/freertos/FreeRTOS-Kernel/portable/port_systick.c b/components/freertos/FreeRTOS-Kernel/portable/port_systick.c
index 9031ef81bd..9c2c7e8e2c 100644
--- a/components/freertos/FreeRTOS-Kernel/portable/port_systick.c
+++ b/components/freertos/FreeRTOS-Kernel/portable/port_systick.c
@@ -94,6 +94,8 @@ void vPortSetupTimer(void)
 
         for (cpuid = 0; cpuid < SOC_CPU_CORES_NUM; cpuid++) {
             systimer_hal_counter_can_stall_by_cpu(&systimer_hal, SYSTIMER_LL_COUNTER_OS_TICK, cpuid, false);
+            uint32_t alarm_id = SYSTIMER_LL_ALARM_OS_TICK_CORE0 + cpuid;
+            systimer_hal_select_alarm_mode(&systimer_hal, alarm_id, SYSTIMER_ALARM_MODE_ONESHOT);
         }
 
         for (cpuid = 0; cpuid < portNUM_PROCESSORS; ++cpuid) {
