diff --git a/components/esp_rom/esp32c2/ld/esp32c2.rom.ld b/components/esp_rom/esp32c2/ld/esp32c2.rom.ld
index 707633b6b0..84a3665eb9 100644
--- a/components/esp_rom/esp32c2/ld/esp32c2.rom.ld
+++ b/components/esp_rom/esp32c2/ld/esp32c2.rom.ld
@@ -1774,7 +1774,7 @@ rcUpdatePhyMode = 0x40001ea8;
 rcGetHighestRateIdx = 0x40001eac;
 pm_tx_null_data_done_process = 0x40001eb0;
 pm_tx_data_process = 0x40001eb4;
-pm_attach = 0x40001eb8;
+/* pm_attach = 0x40001eb8; */
 pm_coex_schm_process = 0x40001ebc;
 ppInitTxq = 0x40001ec0;
 pp_attach = 0x40001ec4;
@@ -1954,7 +1954,7 @@ ieee80211_setup_phy_mode = 0x4000207c;
 ieee80211_sta_is_connected = 0x40002080;
 current_task_is_wifi_task = 0x40002084;
 wifi_get_init_state = 0x40002088;
-ieee80211_timer_process = 0x4000208c;
+/* ieee80211_timer_process = 0x4000208c; */
 cnx_coexist_timeout = 0x40002090;
 /* sta_recv_mgmt = 0x40002094;*/
 ieee80211_send_setup = 0x40002098;
