diff --git a/components/esp_netif/lwip/esp_netif_lwip.c b/components/esp_netif/lwip/esp_netif_lwip.c
index 81aba2a988..beb19f5547 100644
--- a/components/esp_netif/lwip/esp_netif_lwip.c
+++ b/components/esp_netif/lwip/esp_netif_lwip.c
@@ -2264,7 +2264,7 @@ esp_err_t esp_netif_dhcps_option_api(esp_netif_api_msg_t *msg)
                         return ESP_ERR_ESP_NETIF_INVALID_PARAMS;
                     }
                 } else {
-                    return ESP_ERR_ESP_NETIF_INVALID_PARAMS;
+                    // return ESP_ERR_ESP_NETIF_INVALID_PARAMS;
                 }
 
                 memcpy(opt_info, opt->val, opt->len);
