diff --git a/components/bt/host/bluedroid/stack/l2cap/l2c_ble.c b/components/bt/host/bluedroid/stack/l2cap/l2c_ble.c
index 4e6c8534ec..fa7a877491 100644
--- a/components/bt/host/bluedroid/stack/l2cap/l2c_ble.c
+++ b/components/bt/host/bluedroid/stack/l2cap/l2c_ble.c
@@ -950,16 +950,19 @@ BOOLEAN l2cble_init_direct_conn (tL2C_LCB *p_lcb)
         L2CAP_TRACE_ERROR("initate direct connection fail, topology limitation");
         return FALSE;
     }
-    uint32_t link_timeout = L2CAP_BLE_LINK_CONNECT_TOUT;
+
+    extern int32_t bluedroid_conn_wait_time;
+
+    uint32_t link_timeout = bluedroid_conn_wait_time;
     if(GATTC_CONNECT_RETRY_COUNT) {
         if(!p_lcb->retry_create_con) {
             p_lcb->start_time_s = (esp_system_get_time()/1000);
         }
         uint32_t current_time = (esp_system_get_time()/1000);
-        link_timeout = (L2CAP_BLE_LINK_CONNECT_TOUT*1000 - (current_time - p_lcb->start_time_s))/1000;
+        link_timeout = (bluedroid_conn_wait_time*1000 - (current_time - p_lcb->start_time_s))/1000;
 
-        if(link_timeout == 0 || link_timeout > L2CAP_BLE_LINK_CONNECT_TOUT) {
-            link_timeout = L2CAP_BLE_LINK_CONNECT_TOUT;
+        if(link_timeout == 0 || link_timeout > bluedroid_conn_wait_time) {
+            link_timeout = bluedroid_conn_wait_time;
         }
     }
 
